sqldatabaseproductserver/main.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		fmt.Println("error pinging database")
+		return
+	}
+	r := mux.NewRouter()
+	r.HandleFunc("/products", GetProducts).Methods("GET")
+	r.HandleFunc("/products/{id}", GetProduct).Methods("GET")
+	r.HandleFunc("/products", AddProduct).Methods("POST")
+	r.HandleFunc("/products/{id}", UpdateProduct).Methods("PUT")
+	r.HandleFunc("/products/{id}", DeleteProduct).Methods("DELETE")
+	fmt.Println("Server is running")
+	log.Fatal(http.ListenAndServe("localhost:8080", r))
+
+}
+
+func GetProducts(w http.ResponseWriter, r *http.Request) {
+	db, err := sql.Open("sqlite3", "products.db")
+	if err != nil {
+		return
+	}
+	limit := r.FormValue("limit")
+	query := "SELECT * FROM products LIMIT " + limit + ";"
+	if limit == "" {
+		query = "SELECT * FROM products;"
+	}
+	result, err := db.Query(query)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	defer result.Close()
+	for result.Next() {
+		var product Product
+		err := result.Scan(&product.ID, &product.Name, &product.Price)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		products = append(products, product)
+	}
+
+	err = json.NewEncoder(w).Encode(products)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+func GetProduct(w http.ResponseWriter, r *http.Request) {
+	db, err := sql.Open("sqlite3", "products.db")
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	vars := mux.Vars(r)
+	stringid := vars["id"]
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	intid, err := strconv.Atoi(stringid)
+	if intid > 100 {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	query := "SELECT * FROM products WHERE ID = " + stringid + ";"
+	result, err := db.Query(query)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	defer result.Close()
+	var product Product
+	for result.Next() {
+		err := result.Scan(&product.ID, &product.Name, &product.Price)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+	response, err := json.Marshal(product)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	_, err = w.Write(response)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+}
+func AddProduct(w http.ResponseWriter, r *http.Request) {
+	db, err := sql.Open("sqlite3", "products.db")
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	request, err := io.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	newproduct := Product{
+		ID:    0,
+		Name:  "",
+		Price: 0,
+	}
+
+	err = json.Unmarshal(request, &newproduct)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	query := "INSERT INTO products (ID, Name, Price) VALUES (" + string(newproduct.ID) + ",`" + newproduct.Name + "`," + string(newproduct.Price) + ");"
```

Comment:
Your ID is a string?

Reply:
everything in the database query needs to be a string and ID is actually an int so I had to convert it here.

Reply:
Use a prepared statement.